libbpf-tools/fsslower.c

```diff
@@ -0,0 +1,458 @@
+const char argp_program_doc[] =
+"Trace file system operations slower than a threshold.\n"
+"\n"
+"Usage: fsslower [-h] [-t FS] [-p PID] [-m MIN] [-d DURATION] [-c]\n"
+"\n"
+"EXAMPLES:\n"
+"    fsslower -t ext4             # trace ext4 operations slower than 10 ms\n"
+"    fsslower -t nfs -p 1216      # trace nfs operations with PID 1216 only\n"
+"    fsslower -t xfs -c -d 1      # trace xfs operations for 1s with csv output\n";
+
+static const struct argp_option opts[] = {
+	{ "csv", 'c', NULL, 0, "Output as csv" },
+	{ "duration", 'd', "DURATION", 0, "Total duration of trace in seconds" },
+	{ "pid", 'p', "PID", 0, "Process ID to trace" },
+	{ "min", 'm', "MIN", 0, "Min latency to trace, in ms (default 10)" },
+	{ "type", 't', "Filesystem", 0, "Which filesystem to trace, [btrfs/ext4/nfs/xfs]" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 'v':
+		verbose = true;
+		break;
+	case 'c':
+		csv = true;
+		break;
+	case 'd':
+		errno = 0;
+		duration = strtol(arg, NULL, 10);
+		if (errno || duration <= 0) {
+			warn("invalid DURATION: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'm':
+		errno = 0;
+		min_lat_ms = strtoll(arg, NULL, 10);
+		if (errno || min_lat_ms < 0) {
+			warn("invalid latency (in ms): %s\n", arg);
+		}
+		break;
+	case 't':
+		if (!strcmp(arg, "btrfs")) {
+			fs_type = BTRFS;
+		} else if (!strcmp(arg, "ext4")) {
+			fs_type = EXT4;
+		} else if (!strcmp(arg, "nfs")) {
+			fs_type = NFS;
+		} else if (!strcmp(arg, "xfs")) {
+			fs_type = XFS;
+		} else {
+			warn("invalid filesystem\n");
+			argp_usage(state);
+		}
+		break;
+	case 'p':
+		errno = 0;
+		target_pid = strtol(arg, NULL, 10);
+		if (errno || target_pid <= 0) {
+			warn("invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static void alias_parse(char *prog)
+{
+	char *name = basename(prog);
+
+	if (!strcmp(name, "btrfsslower")) {
+		fs_type = BTRFS;
+	} else if (!strcmp(name, "ext4slower")) {
+		fs_type = EXT4;
+	} else if (!strcmp(name, "nfsslower")) {
+		fs_type = NFS;
+	} else if (!strcmp(name, "xfsslower")) {
+		fs_type = XFS;
+	}
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level,
+			   const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = 1;
+}
+
+static bool check_fentry()
+{
+	int i;
+	const char *fn_name, *module;
+	bool support_fentry = true;
+
+	for (i = 0; i < MAX_OP; i++) {
+		fn_name = fs_configs[fs_type].op_funcs[i];
+		module = fs_configs[fs_type].fs;
+		if (fn_name && !fentry_exists(fn_name, NULL)
+		    && !fentry_exists(fn_name, module)) {
```

Comment:
hm... this made me realize that fentry_exists behavior is suboptimal. I think it should treat module as a hint, not as a mandatory thing. I.e., if we do fentry_exists("my_func", "my_module") and there is no /sys/kernel/btf/my_module, then just ignore "my_module" and try to find the type in the vmlinux. So that you can check existence of a function with just one fentry_exists() call. If functionality is in a module and that module exists -- good, we know which module to look at, but if it was compiled in or module BTFs are not supported, still look into vmlinux BTF.

In either case, such behavior will return true only if we did find the function in BTF somewhere. Can you please fix fentry_exists() to behave like that?

Reply:
Done.

Reply:
if it's done, you should remove `fentry_exists(fn_name, NULL)`